analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -75,16 +75,19 @@ public override ProgramState ConditionEvaluated(SymbolicContext context)
     private bool IsIgnored(ProgramState state, IOperation operation) =>
         operation.Kind is OperationKindEx.Literal
         || IsVarPattern(operation)
-        || IsDiscardPattern(operation)
+        || IsDiscardPattern(state, operation)
         || operation.TrackedSymbol(state) is IFieldSymbol { IsConst: true }
         || operation.Syntax.Ancestors().Any(x => IsInsideUsingDeclaration(x) || IsLockStatement(x));
 
     private static bool IsVarPattern(IOperation operation) =>
         operation.AsIsPattern()?.Pattern.WrappedOperation.AsDeclarationPattern()?.MatchesNull is true;
 
-    private static bool IsDiscardPattern(IOperation operation) =>
-        operation.AsIsPattern() is { } pattern
-        && pattern.Pattern.WrappedOperation.Kind is OperationKindEx.DiscardPattern;
+    private static bool IsDiscardPattern(ProgramState state, IOperation operation) =>
+        operation.Kind is OperationKindEx.DiscardPattern
+        || (operation.AsIsPattern() is { } isPattern
+            && state.ResolveCaptureAndUnwrapConversion(isPattern.Pattern.WrappedOperation) is var subPattern
```

Comment:
I could not find a case where `ResolveCaptureAndUnwrapConversion` would be necessary.

Reply:
You're right. I misread [the CFG diagram](https://edotor.net/?engine=dot#%2F%2A%0D%0Ausing%20System.Collections.Generic%3B%0D%0A%0D%0Aint%20Compare%28string%20name1%2C%20string%20name2%29%20%3D%3E%20%28%28name1%2C%20name1%29%2C%20name2%29%20switch%0D%0A%7B%0D%0A%20%20%20%20%28%28null%2C%20null%29%2C%20null%29%20%3D%3E%200%2C%0D%0A%20%20%20%20%28%28null%2C%20null%29%2C%20_%29%20%3D%3E%201%2C%0D%0A%20%20%20%20%28%28_%2C%20_%29%2C%20null%29%20%3D%3E%20-1%2C%0D%0A%20%20%20%20%28%28_%2C%20_%29%2C%20_%29%20%3D%3E%20Comparer%3Cstring%3E.Default.Compare%28name1%2C%20name2%29%0D%0A%7D%3B%0D%0A%0D%0Aclass%20Person%0D%0A%7B%0D%0A%20%20%20%20public%20string%20Name%20%7B%20get%3B%20set%3B%20%7D%0D%0A%7D%0D%0A%2A%2F%0D%0A%0D%0Adigraph%20%22RoslynCFG%22%20%7B%0D%0Asubgraph%20%22cluster_1%22%20%7B%0Alabel%20%3D%20%22LocalLifetime%20region%22%0D%0Acfg0_block1%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%231%7D%22%5D%0D%0A%7D%0D%0Acfg0_block0%20%5Bshape%3Drecord%20label%3D%22%7BENTRY%20%230%7D%22%5D%0D%0Acfg0_block2%20%5Bshape%3Drecord%20label%3D%22%7BEXIT%20%232%7D%22%5D%0D%0Asubgraph%20%22cluster_3%22%20%7B%0Alabel%20%3D%20%22RoslynCFG.Compare%22%0D%0Asubgraph%20%22cluster_4%22%20%7B%0Alabel%20%3D%20%22LocalLifetime%20region%2C%20Captures%3A%20%23Capture-0%22%0D%0Asubgraph%20%22cluster_5%22%20%7B%0Alabel%20%3D%20%22LocalLifetime%20region%2C%20Captures%3A%20%23Capture-1%22%0D%0Acfg2_block1%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%231%7C0%23%3A%20FlowCaptureOperation%3A%20%23Capture-1%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%7C1%23%3A%200%23.Value%3A%20TupleOperation%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%7C2%23%3A%20TupleOperation%3A%20%28name1%2C%20name1%29%7C3%23%3A%20ParameterReferenceOperation%3A%20name1%7C3%23%3A%20ParameterReferenceOperation%3A%20name1%7C2%23%3A%20ParameterReferenceOperation%3A%20name2%7C%23%23%23%23%23%23%23%23%23%23%7C%23%23%20BranchValue%20%23%23%7C0%23%3A%20IsPatternOperation%3A%20%28%28null%2C%20null%29%2C%20null%29%20%3D%5C%3E%200%7C1%23%3A%200%23.Value%3A%20FlowCaptureReferenceOperation%3A%20%23CaptureRef-1%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%7C1%23%3A%200%23.Pattern%3A%20RecursivePatternOperation%3A%20%28%28null%2C%20null%29%2C%20null%29%7C2%23%3A%20RecursivePatternOperation%3A%20%28null%2C%20null%29%7C3%23%3A%20ConstantPatternOperation%3A%20null%7C4%23%3A%203%23.Value%3A%20ConversionOperation%3A%20null%7C5%23%3A%204%23.Operand%3A%20LiteralOperation%3A%20null%7C3%23%3A%20ConstantPatternOperation%3A%20null%7C4%23%3A%203%23.Value%3A%20ConversionOperation%3A%20null%7C5%23%3A%204%23.Operand%3A%20LiteralOperation%3A%20null%7C2%23%3A%20ConstantPatternOperation%3A%20null%7C3%23%3A%202%23.Value%3A%20ConversionOperation%3A%20null%7C4%23%3A%203%23.Operand%3A%20LiteralOperation%3A%20null%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block2%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%232%7C0%23%3A%20FlowCaptureOperation%3A%20%23Capture-0%3A%200%7C1%23%3A%200%23.Value%3A%20LiteralOperation%3A%200%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block3%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%233%7C%23%23%20BranchValue%20%23%23%7C0%23%3A%20IsPatternOperation%3A%20%28%28null%2C%20null%29%2C%20_%29%20%3D%5C%3E%201%7C1%23%3A%200%23.Value%3A%20FlowCaptureReferenceOperation%3A%20%23CaptureRef-1%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%7C1%23%3A%200%23.Pattern%3A%20RecursivePatternOperation%3A%20%28%28null%2C%20null%29%2C%20_%29%7C2%23%3A%20RecursivePatternOperation%3A%20%28null%2C%20null%29%7C3%23%3A%20ConstantPatternOperation%3A%20null%7C4%23%3A%203%23.Value%3A%20ConversionOperation%3A%20null%7C5%23%3A%204%23.Operand%3A%20LiteralOperation%3A%20null%7C3%23%3A%20ConstantPatternOperation%3A%20null%7C4%23%3A%203%23.Value%3A%20ConversionOperation%3A%20null%7C5%23%3A%204%23.Operand%3A%20LiteralOperation%3A%20null%7C2%23%3A%20DiscardPatternOperation%3A%20_%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block4%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%234%7C0%23%3A%20FlowCaptureOperation%3A%20%23Capture-0%3A%201%7C1%23%3A%200%23.Value%3A%20LiteralOperation%3A%201%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block5%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%235%7C%23%23%20BranchValue%20%23%23%7C0%23%3A%20IsPatternOperation%3A%20%28%28_%2C%20_%29%2C%20null%29%20%3D%5C%3E%20-1%7C1%23%3A%200%23.Value%3A%20FlowCaptureReferenceOperation%3A%20%23CaptureRef-1%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%7C1%23%3A%200%23.Pattern%3A%20RecursivePatternOperation%3A%20%28%28_%2C%20_%29%2C%20null%29%7C2%23%3A%20RecursivePatternOperation%3A%20%28_%2C%20_%29%7C3%23%3A%20DiscardPatternOperation%3A%20_%7C3%23%3A%20DiscardPatternOperation%3A%20_%7C2%23%3A%20ConstantPatternOperation%3A%20null%7C3%23%3A%202%23.Value%3A%20ConversionOperation%3A%20null%7C4%23%3A%203%23.Operand%3A%20LiteralOperation%3A%20null%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block6%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%236%7C0%23%3A%20FlowCaptureOperation%3A%20%23Capture-0%3A%20-1%7C1%23%3A%200%23.Value%3A%20UnaryOperation%3A%20-1%7C2%23%3A%201%23.Operand%3A%20LiteralOperation%3A%201%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block7%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%237%7C%23%23%20BranchValue%20%23%23%7C0%23%3A%20IsPatternOperation%3A%20%28%28_%2C%20_%29%2C%20_%29%20%3D%5C%3E%20Comparer%5C%3Cstring%5C%3E.Default.Compare%28name1%2C%20name2%29%7C1%23%3A%200%23.Value%3A%20FlowCaptureReferenceOperation%3A%20%23CaptureRef-1%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%7C1%23%3A%200%23.Pattern%3A%20RecursivePatternOperation%3A%20%28%28_%2C%20_%29%2C%20_%29%7C2%23%3A%20RecursivePatternOperation%3A%20%28_%2C%20_%29%7C3%23%3A%20DiscardPatternOperation%3A%20_%7C3%23%3A%20DiscardPatternOperation%3A%20_%7C2%23%3A%20DiscardPatternOperation%3A%20_%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block8%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%238%7C0%23%3A%20FlowCaptureOperation%3A%20%23Capture-0%3A%20Comparer%5C%3Cstring%5C%3E.Default.Compare%28name1%2C%20name2%29%7C1%23%3A%200%23.Value%3A%20InvocationOperation%3A%20Compare%3A%20Comparer%5C%3Cstring%5C%3E.Default.Compare%28name1%2C%20name2%29%7C2%23%3A%201%23.Instance%3A%20PropertyReferenceOperation%3A%20Comparer%5C%3Cstring%5C%3E.Default%7C2%23%3A%20ArgumentOperation%3A%20name1%7C3%23%3A%202%23.Value%3A%20ParameterReferenceOperation%3A%20name1%7C2%23%3A%20ArgumentOperation%3A%20name2%7C3%23%3A%202%23.Value%3A%20ParameterReferenceOperation%3A%20name2%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0A%7D%0D%0Acfg2_block9%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%239%7C%23%23%20BranchValue%20%23%23%7C0%23%3A%20ObjectCreationOperation%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%20switch%5Cn%5C%7B%5Cn%20%20%20%20%28%28null%2C%20null%29%2C%20null%29%20%3D%5C%3E%200%2C%5Cn%20%20%20%20%28%28null%2C%20null%29%2C%20_%29%20%3D%5C%3E%201%2C%5Cn%20%20%20%20%28%28_%2C%20_%29%2C%20null%29%20%3D%5C%3E%20-1%2C%5Cn%20%20%20%20%28%28_%2C%20_%29%2C%20_%29%20%3D%5C%3E%20Comparer%5C%3Cstring%5C%3E.Default.Compare%28name1%2C%20name2%29%5Cn%5C%7D%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0Acfg2_block10%20%5Bshape%3Drecord%20label%3D%22%7BBLOCK%20%2310%7C%23%23%20BranchValue%20%23%23%7C0%23%3A%20FlowCaptureReferenceOperation%3A%20%23CaptureRef-0%3A%20%28%28name1%2C%20name1%29%2C%20name2%29%20switch%5Cn%5C%7B%5Cn%20%20%20%20%28%28null%2C%20null%29%2C%20null%29%20%3D%5C%3E%200%2C%5Cn%20%20%20%20%28%28null%2C%20null%29%2C%20_%29%20%3D%5C%3E%201%2C%5Cn%20%20%20%20%28%28_%2C%20_%29%2C%20null%29%20%3D%5C%3E%20-1%2C%5Cn%20%20%20%20%28%28_%2C%20_%29%2C%20_%29%20%3D%5C%3E%20Comparer%5C%3Cstring%5C%3E.Default.Compare%28name1%2C%20name2%29%5Cn%5C%7D%7C%23%23%23%23%23%23%23%23%23%23%7D%22%5D%0D%0A%7D%0D%0Acfg2_block0%20%5Bshape%3Drecord%20label%3D%22%7BENTRY%20%230%7D%22%5D%0D%0Acfg2_block11%20%5Bshape%3Drecord%20label%3D%22%7BEXIT%20%2311%7D%22%5D%0D%0A%7D%0D%0Acfg0_block0%20-%3E%20cfg0_block1%0D%0Acfg0_block1%20-%3E%20cfg0_block2%0D%0Acfg2_block0%20-%3E%20cfg2_block1%0D%0Acfg2_block1%20-%3E%20cfg2_block2%20%5Blabel%3D%22Else%22%5D%0D%0Acfg2_block1%20-%3E%20cfg2_block3%20%5Blabel%3D%22WhenFalse%22%5D%0D%0Acfg2_block3%20-%3E%20cfg2_block4%20%5Blabel%3D%22Else%22%5D%0D%0Acfg2_block3%20-%3E%20cfg2_block5%20%5Blabel%3D%22WhenFalse%22%5D%0D%0Acfg2_block5%20-%3E%20cfg2_block6%20%5Blabel%3D%22Else%22%5D%0D%0Acfg2_block5%20-%3E%20cfg2_block7%20%5Blabel%3D%22WhenFalse%22%5D%0D%0Acfg2_block7%20-%3E%20cfg2_block8%20%5Blabel%3D%22Else%22%5D%0D%0Acfg2_block7%20-%3E%20cfg2_block9%20%5Blabel%3D%22WhenFalse%22%5D%0D%0Acfg2_block9%20-%3E%20NoDestination_cfg2_block9%20%5Blabel%3D%22Throw%22%5D%0D%0Acfg2_block2%20-%3E%20cfg2_block10%0D%0Acfg2_block4%20-%3E%20cfg2_block10%0D%0Acfg2_block6%20-%3E%20cfg2_block10%0D%0Acfg2_block8%20-%3E%20cfg2_block10%0D%0Acfg2_block10%20-%3E%20cfg2_block11%20%5Blabel%3D%22Return%22%5D%0D%0A%7D%0D%0A) and looked the `Value` property of the operation instead of `Pattern`.